File: engines/typescript/src/catalog.ts
/**
 * Load a control catalog (SPEC §7.1, §7.3-7.4).
 *
 * A catalog is a directory with `catalog.yaml` (metadata and the control list), `controls/*.yaml` (one
 * control each), and `shapes/*.ttl` (the Portable Shape Profile shapes the rung-2 controls reference).
 * This is a faithful port of the reference loader; catalog linting (which re-runs every fixture through
 * the evaluator) is not needed by the conformance pipeline and is not ported here.
 */

import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { load } from "js-yaml";
import { type Shape, loadShapes } from "./psp";
import { byteCompare } from "./util";

export interface ControlSpec {
  id: string;
  version: string;
  title: string;
  appliesToRoles: string[];
  mode: string;
  rung: number;
  severity: string;
  minSourceClass: string;
  minimumEvidence: Array<Record<string, string>>;
  shapePath: string | null;
  tolerance: Record<string, unknown>;
  testCases: Array<Record<string, string>>;
  raw: Record<string, unknown>;
}

export interface Catalog {
  id: string;
  version: string;
  directory: string;
  controls: ControlSpec[];
  shapes: Map<string, Shape>;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function yamlFiles(directory: string): string[] {
  try {
    return readdirSync(directory)
      .filter((name) => name.endsWith(".yaml"))
      .sort(byteCompare);
  } catch {
    return [];
  }
}

function controlFromDict(data: Record<string, unknown>): ControlSpec {
  const evaluation = isRecord(data.evaluation) ? data.evaluation : {};
  const applicability = isRecord(data.applicability) ? data.applicability : {};
  return {
    id: String(data.id ?? ""),
    version: String(data.version ?? ""),
    title: String(data.title ?? ""),
    appliesToRoles: Array.isArray(applicability.applies_to_roles)
      ? applicability.applies_to_roles.map(String)
      : [],
    mode: String(evaluation.mode ?? ""),
    rung: Number.parseInt(String(evaluation.rung ?? 0), 10),
    severity: String(data.severity ?? ""),
    minSourceClass: String(evaluation.min_source_class ?? "any"),
    minimumEvidence: Array.isArray(evaluation.minimum_evidence)
      ? (evaluation.minimum_evidence as Array<Record<string, string>>)
      : [],
    shapePath: typeof evaluation.shape === "string" ? evaluation.shape : null,
    tolerance: isRecord(data.tolerance) ? data.tolerance : { kind: "count", max: 0 },
    testCases: Array.isArray(data.test_cases)
      ? (data.test_cases as Array<Record<string, string>>)
      : [],
    raw: data,
  };
}

/** Load `catalog.yaml` and every control and shape in `directory`. */
export function loadCatalog(directory: string): Catalog {
  const meta = load(readFileSync(join(directory, "catalog.yaml"), "utf-8"));
  const metaRecord = isRecord(meta) ? meta : {};
  const controls: ControlSpec[] = [];
  const shapes = new Map<string, Shape>();
  for (const controlFile of yamlFiles(join(directory, "controls"))) {
    const data = load(readFileSync(join(directory, "controls", controlFile), "utf-8"));
    if (!isRecord(data)) {
      continue;
    }
    const control = controlFromDict(data);
    controls.push(control);
    if (control.shapePath) {
      for (const [iri, shape] of loadShapes(join(directory, control.shapePath))) {
        shapes.set(iri, shape);
      }
    }
  }
  return {
    id: String(metaRecord.id ?? ""),
    version: String(metaRecord.version ?? ""),
    directory,
    controls,
    shapes,
  };
}

/** The PSP shape a control evaluates against (by `<id>-Shape` suffix, then any shape with targets). */
export function shapeFor(catalog: Catalog, control: ControlSpec): Shape | null {
  if (control.shapePath === null) {
    return null;
  }
  const want = `${control.id}-Shape`;
  for (const [iri, shape] of catalog.shapes) {
    if (iri.endsWith(want)) {
      return shape;
    }
  }
  for (const shape of catalog.shapes.values()) {
    if (shape.targetClass || shape.targetNodes.length > 0 || shape.targetWhere.length > 0) {
      return shape;
    }
  }
  return null;
}

File: engines/java/src/main/java/org/agentce/Catalog.java
package org.agentce;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

/**
 * Load a control catalog (SPEC §7.1, §7.3-7.4): a directory with {@code catalog.yaml} (metadata and the
 * control list), {@code controls/*.yaml} (one control each), and {@code shapes/*.ttl} (the Portable
 * Shape Profile shapes the rung-2 controls reference). A faithful port of the reference loader.
 */
public final class Catalog {
    public final String id;
    public final String version;
    public final Path directory;
    public final List<ControlSpec> controls;
    public final Map<String, Psp.Shape> shapes;

    private Catalog(String id, String version, Path directory, List<ControlSpec> controls, Map<String, Psp.Shape> shapes) {
        this.id = id;
        this.version = version;
        this.directory = directory;
        this.controls = controls;
        this.shapes = shapes;
    }

    public static final class ControlSpec {
        public String id;
        public String version;
        public String title;
        public List<String> appliesToRoles = new ArrayList<>();
        public String mode;
        public int rung;
        public String severity;
        public String minSourceClass;
        public List<JsonNode> minimumEvidence = new ArrayList<>();
        public String shapePath;
        public JsonNode tolerance;
        public List<JsonNode> testCases = new ArrayList<>();
        public JsonNode raw;
    }

    private static String text(JsonNode node, String field, String dflt) {
        JsonNode v = node.get(field);
        return v != null && !v.isNull() ? v.asText() : dflt;
    }

    private static ControlSpec controlFromDict(JsonNode data) {
        JsonNode evaluation = data.get("evaluation");
        if (evaluation == null || !evaluation.isObject()) {
            evaluation = Json.nodes().objectNode();
        }
        JsonNode applicability = data.get("applicability");
        if (applicability == null || !applicability.isObject()) {
            applicability = Json.nodes().objectNode();
        }
        ControlSpec control = new ControlSpec();
        control.id = text(data, "id", "");
        control.version = text(data, "version", "");
        control.title = text(data, "title", "");
        JsonNode roles = applicability.get("applies_to_roles");
        if (roles != null && roles.isArray()) {
            for (JsonNode r : roles) {
                control.appliesToRoles.add(r.asText());
            }
        }
        control.mode = text(evaluation, "mode", "");
        control.rung = Integer.parseInt(text(evaluation, "rung", "0"));
        control.severity = text(data, "severity", "");
        control.minSourceClass = text(evaluation, "min_source_class", "any");
        JsonNode minEvidence = evaluation.get("minimum_evidence");
        if (minEvidence != null && minEvidence.isArray()) {
            minEvidence.forEach(control.minimumEvidence::add);
        }
        JsonNode shape = evaluation.get("shape");
        control.shapePath = shape != null && shape.isTextual() ? shape.textValue() : null;
        JsonNode tolerance = data.get("tolerance");
        if (tolerance != null && tolerance.isObject()) {
            control.tolerance = tolerance;
        } else {
            ObjectNode dflt = Json.nodes().objectNode();
            dflt.put("kind", "count");
            dflt.put("max", 0);
            control.tolerance = dflt;
        }
        JsonNode testCases = data.get("test_cases");
        if (testCases != null && testCases.isArray()) {
            testCases.forEach(control.testCases::add);
        }
        control.raw = data;
        return control;
    }

    private static List<String> yamlFiles(Path directory) {
        List<String> names = new ArrayList<>();
        if (!Files.isDirectory(directory)) {
            return names;
        }
        try (Stream<Path> stream = Files.list(directory)) {
            stream.map(p -> p.getFileName().toString())
                    .filter(n -> n.endsWith(".yaml"))
                    .forEach(names::add);
        } catch (IOException e) {
            return names;
        }
        names.sort(Json::byteCompare);
        return names;
    }

    /** Load {@code catalog.yaml} and every control and shape in {@code directory}. */
    public static Catalog load(Path directory) {
        JsonNode meta = Yaml.parseFile(directory.resolve("catalog.yaml"));
        JsonNode metaRecord = meta != null && meta.isObject() ? meta : Json.nodes().objectNode();
        List<ControlSpec> controls = new ArrayList<>();
        Map<String, Psp.Shape> shapes = new LinkedHashMap<>();
        for (String controlFile : yamlFiles(directory.resolve("controls"))) {
            JsonNode data = Yaml.parseFile(directory.resolve("controls").resolve(controlFile));
            if (data == null || !data.isObject()) {
                continue;
            }
            ControlSpec control = controlFromDict(data);
            controls.add(control);
            if (control.shapePath != null) {
                shapes.putAll(Psp.loadShapes(directory.resolve(control.shapePath)));
            }
        }
        return new Catalog(
                text(metaRecord, "id", ""), text(metaRecord, "version", ""), directory, controls, shapes);
    }

    /** The PSP shape a control evaluates against (by {@code <id>-Shape} suffix, then any shape with targets). */
    public static Psp.Shape shapeFor(Catalog catalog, ControlSpec control) {
        if (control.shapePath == null) {
            return null;
        }
        String want = control.id + "-Shape";
        for (Map.Entry<String, Psp.Shape> e : catalog.shapes.entrySet()) {
            if (e.getKey().endsWith(want)) {
                return e.getValue();
            }
        }
        for (Psp.Shape shape : catalog.shapes.values()) {
            if (shape.targetClass != null || !shape.targetNodes.isEmpty() || !shape.targetWhere.isEmpty()) {
                return shape;
            }
        }
        return null;
    }
}
